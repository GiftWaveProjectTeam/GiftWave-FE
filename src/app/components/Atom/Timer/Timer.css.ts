import { vars } from "@/app/styles/themes.css";
import { styleVariants } from "@vanilla-extract/css";

export const TimerVariant = styleVariants({
  seconds: {
    fontSize: "14px",
    width: "3.6em",
    color: "gray",
  },
  minutes: {
    fontSize: vars.text.R_16.fontSize,
    fontWeight: vars.text.R_16.fontWeight,
    color: vars.palettle.blue,
    width: "3.6em",
  },
  hours: {
    fontSize: "14px",
    width: "5.5em",
    color: "gray",
  },
});
