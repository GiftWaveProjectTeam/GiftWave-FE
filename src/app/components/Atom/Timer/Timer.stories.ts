import { Meta, StoryObj } from "@storybook/react";
import Timer from "./Timer";

const meta: Meta<typeof Timer> = {
  title: "Atom/Timer",
  component: Timer,
};

export default meta;
type Story = StoryObj<typeof Timer>;

export const TimerBox: Story = {
  args: {},
};
