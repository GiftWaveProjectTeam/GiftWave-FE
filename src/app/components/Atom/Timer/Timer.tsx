"use client";

import { useEffect, useState } from "react";
import React from "react";
import { TimerVariant } from "./Timer.css";
import { useInterval } from "@hooks";
import classNames from "classnames";

interface TimerProps {
  maxTime: number;
  mode?: "seconds" | "minutes" | "hours";
  trigger?: boolean;
  setTrigger?: React.Dispatch<React.SetStateAction<boolean>>;
  timeout?: React.Dispatch<React.SetStateAction<boolean>>;
}

const Timer = ({
  maxTime,
  trigger,
  setTrigger,
  timeout,
  mode = "minutes",
}: TimerProps) => {
  const [timer, setTimer] = useState<number>(maxTime);

  const hours = Math.floor(timer / 3600)
    .toString()
    .padStart(2, "0");
  const minutes = Math.floor(timer / 60)
    .toString()
    .padStart(2, "0");
  const seconds = (timer % 60).toString().padStart(2, "0");

  timeout && timer === 0 && timeout(true);

  if (timer === 0) {
    setTrigger && setTrigger(false);
  }
  useInterval(
    () => {
      setTimer(timer - 1);
    },
    1000,
    trigger
  );

  useEffect(() => {
    setTimer(maxTime);
  }, [maxTime]);

  return (
    <div className={classNames(TimerVariant[mode])}>
      {mode === "hours" &&
        (hours !== "0" ? <span>{hours}:</span> : <span>00</span>)}
      {minutes !== "0" ? <span>{minutes}:</span> : <span>00:</span>}
      {seconds !== "0" ? <span>{seconds}</span> : <span>00</span>}
    </div>
  );
};

export default Timer;
