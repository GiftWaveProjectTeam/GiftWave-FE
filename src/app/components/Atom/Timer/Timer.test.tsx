import React from "react";
import { render, waitFor } from "@testing-library/react";
import Timer from "./Timer";
import { act } from "react-dom/test-utils";

describe("Timer component", () => {
  it("renders correctly with default mode", () => {
    const { container } = render(<Timer maxTime={90} trigger={false} />);
    expect(container).toMatchSnapshot();
  });

  it("renders correctly with hours mode", () => {
    const { container } = render(
      <Timer maxTime={3600} trigger={false} mode="hours" />
    );
    expect(container).toMatchSnapshot();
  });

  it("decreases timer and formats time correctly", async () => {
    jest.useFakeTimers();
    const { container } = render(<Timer maxTime={10} trigger={true} />);

    const expectedTime =
      "<div><div><span>00 : </span><span>10</span></div></div>";
    expect(container).toContainHTML(expectedTime);

    await act(async () => {
      jest.advanceTimersByTime(1000);
    });

    const expectedTime2 =
      "<div><div><span>00 : </span><span>09</span></div></div>";

    await waitFor(() => expect(container).toContainHTML(expectedTime2));

    await act(async () => {
      jest.advanceTimersByTime(1000);
    });

    const expectedTime3 =
      "<div><div><span>00 : </span><span>08</span></div></div>";
    await waitFor(() => expect(container).toContainHTML(expectedTime3));
    jest.useRealTimers();
  });
});
