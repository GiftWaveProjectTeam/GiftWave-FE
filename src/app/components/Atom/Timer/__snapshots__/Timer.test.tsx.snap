// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer component renders correctly with default mode 1`] = `
<div>
  <div>
    <span>
      01
       : 
    </span>
    <span>
      30
    </span>
  </div>
</div>
`;

exports[`Timer component renders correctly with hours mode 1`] = `
<div>
  <div>
    <span>
      01
       : 
    </span>
    <span>
      60
       : 
    </span>
    <span>
      00
    </span>
  </div>
</div>
`;
